카드 오브젝트 마우스 올리면 위로 올라가는 애니메이션

Front-line-dev opened the pull request:
## 💁 설명

카드 오브젝트 클래스의 메소드로 구현하였고 이걸 구현하기 위해서 게임 오브젝트의 메소드 파라미터 타입을 바꿈
position, move
angle
모두 string을 줬었는데 이제부터는 number로 파라미터를 줘야함

사용 방법
Card.move(10,20)
Card.rotate(30)
Card.setAnimateMove()
Card.deleteAnimateMove()

## 📑 체크리스트

> 구현한 목록 체크리스트

- [ ] 카드에 마우스를 올리면 카드가 올라가는 애니메이션

## 🚧 주의 사항

> PR을 읽을 때 살펴볼 사항

쓰로틀 시도해봤으나 실패했습니다

Front-line-dev committed:
refactor: 파라미터를 string에서 int로

this.angle = 0;
this.position = [0, 0];

Front-line-dev committed:
feat: 카드 오브젝트 moveUp 함수

사용 방법
Card.move(10,20)
Card.rotate(30)
Card.setAnimateMove()
Card.deleteAnimateMove()

sbyeol3 commented on src/frontend/engine/CardObject.js:
```diff
@@ -3,34 +3,38 @@ import GameObject from './GameObject';
 import ImageObject from './ImageObject';
 
 const getFacingStyle = (isUp) => (isUp ? 'rotateY(0deg)' : 'rotateY(180deg)');
+const deg2rad = (deg) => (deg * Math.PI) / 180;
```
여기서 2는 `to` 인거죠?ㅋㅋㅋ 갑자기 두 글자 치기 귀찮아지신건가요

Front-line-dev replied:
저는 ~to~ 할때 항상 ~2~로 작성했어요 ㅋㅋ 그냥 제 스타일입니다. 맘에 안 드신다면 바꿔도 돼요.

jinhyukoo replied:
저도 갠적으로 to가 더 좋아보이지만 코드 스타일이니 그냥 두는걸루!

sbyeol3 commented on src/frontend/engine/GameObject.js:
```diff
@@ -87,37 +90,42 @@ const GameObject = class {
   }
 
-  move(x = '0%', y = '0%', duration = 0.5) {
+  move(x = 0, y = 0, duration = 0.5) {
```
duration 초기값도 유틸로 빼서 관리하면 더 좋을듯 합니다. 지금 제가 유틸에 추가한게 잇어서 나중에 제가 바꿔놓을게요~

Front-line-dev replied:
아 감사합니다!

sbyeol3 commented on src/frontend/engine/GameObject.js:
```diff
@@ -87,37 +90,42 @@ const GameObject = class {
-  move(x = '0%', y = '0%', duration = 0.5) {
+  move(x = 0, y = 0, duration = 0.5) {
+    this.position = [x, y];
+    const xString = `${x}%`;
```
이거 자주쓰일거같으면 차라리 메소드 추가해도 갠찬을듯합니다.
addPercent 같은 느낌

Front-line-dev replied:
그러게요 %를 떼는건 replacePercent라는 함수가 위에 있는데 반대로 붙이는 함수를 만들어도 괜찮겠네요. 커플로 만들어보도록 하겠습니다. replacePercent라는 함수가 꼭 %만 떼는게 아니라 딴거도 떼더라고요

sbyeol3 commented on src/frontend/engine/GameObject.js:
```diff
@@ -87,37 +90,42 @@ const GameObject = class {
+
     if (this.animationFrame) cancelAnimationFrame(this.animationFrame);
-    if (!x && !y) {
+    if (!xString && !yString) {
```
위에서 x랑 y가 널이어도 null%가 되기 때문에 if에서 절대 안걸릴 것 같은데욥?

Front-line-dev replied:
원본 코드에서 x를 y로 바꾸다보니 이렇게 됐네요... 고치겠습니다

sbyeol3 commented on src/frontend/engine/GameObject.js:
```diff
@@ -128,19 +136,21 @@ const GameObject = class {
-  rotate(angle = '0deg', duration = 0.2) {
+  rotate(angle = 0, duration = 0.2) {
+    this.angle = angle;
+    const angleString = `${angle}deg`;
```
위에랑 비슷한 로직인데 addUnitToNumber(num, unit) 느낌

Front-line-dev replied:
그런식으로 만들어도 좋을거같네요. 그러면 이렇게 만들어보겠습니다

sbyeol3 reviewed (commented):
굉장히 어려워보이는 코드 덜덜덜 💥

Front-line-dev committed:
feat: moveUp moveDown 함수 받기 가능

moveUp = Card.moveUpCallback()
moveDown = Card.moveDownCallback()

moveUp()
moveDown()

Front-line-dev committed:
feat: CSS card-glow-class

hover하면 빛나는 버전
아무것도 안 해도 빛나는 버전

Front-line-dev commented:
이제 moveUp 함수와 moveDown 함수를 받아서 socket.on 함수 안에서 사용할 수 있습니다. 또한 card-glow css class를 생성했습니다.(임시)

샛별님이 코드리뷰를 아주 잘 해주셔서 감사합니다. 저녁 먹고 반영하고 머지하겠습니다.

ramram1048 commented on src/frontend/engine/CardObject.js:
```diff
@@ -3,34 +3,39 @@ import GameObject from './GameObject';
     imagePath = backImage, // default: 뒷면 이미지
     facingUp = true,
-    hoverable = false,
+    hoverable = false, // <- 안 쓰는게 나을지도?
```
넵 마음대로 하셔도되요 !!

ramram1048 commented on src/frontend/utils/common.scss:
```diff
@@ -83,3 +83,15 @@ html {
+}
+
+.card-glow-hover:hover {
+  box-shadow: 0px 0px 36px 5px rgba(235,222,76,1);
```
`rgba(235,222,76,1)` 이나 `0px 0px 36px 5px rgba(235,222,76,1)` 전체를 scss variable로 정의할 수 있을 것 같습니다!

Front-line-dev replied:
그렇군요! 색을 몇개 지정해놔야겠네요

jinhyukoo reviewed (approved):
샛별님이 리뷰를 굉장히 잘해주셨네요!! 저는 전체 코드를 다시 한 번 읽어봐야겠네요~ 고생 많으셨습니다!!

Front-line-dev committed:
refactor: 리뷰 반영

https://github.com/boostcamp-2020/Project18-B-Web-Duxit/pull/84#discussion_r533181363

Front-line-dev commented:
makeUnitString
makeFloat
이렇게 리팩토링 했습니다

if (!xString && !yString)
이 부분은 move 메소드에 원래 있던 코드라 물어보고 고쳐야될거 같아요

Front-line-dev committed:
refactor: 리뷰반영: scss 색 지정

gold 컬러를 샘플로 만들어보았습니다

sbyeol3 commented:
죻습니다~~
